File: articles-api-typescript/src/services/ArticleService.ts
import { DataSource, In } from "typeorm";
import { Article } from "../entities/Article";
import { Tag } from "../entities/Tag";
import { User } from "../entities/User";
import { FollowRelation } from "../entities/FollowRelation";
import { ArticleFavorite } from "../entities/ArticleFavorite";
import { ArticleResponse, ProfileResponse } from "../types";

export class ArticleService {
  constructor(private dataSource: DataSource) {}

  /**
   * Build a ProfileResponse (matching Java's ProfileData with @JsonIgnore on id).
   */
  private buildProfile(user: User, following: boolean): ProfileResponse {
    return {
      username: user.username,
      bio: user.bio || null,
      image: user.image || null,
      following,
    };
  }

  /**
   * Build an ArticleResponse matching the Java ArticleData shape.
   * tagList is sorted alphabetically.
   */
  private buildArticleResponse(
    article: Article,
    author: User,
    favorited: boolean,
    favoritesCount: number,
    followingAuthor: boolean
  ): ArticleResponse {
    const tagList = article.tags
      .map((t) => t.name)
      .sort((a, b) => a.localeCompare(b));

    return {
      id: article.id,
      slug: article.slug,
      title: article.title,
      description: article.description,
      body: article.body,
      favorited,
      favoritesCount,
      createdAt: article.createdAt.toISOString(),
      updatedAt: article.updatedAt.toISOString(),
      tagList,
      author: this.buildProfile(author, followingAuthor),
    };
  }

  /**
   * Check if currentUser is following the author.
   */
  private async isFollowing(
    currentUserId: string | undefined,
    targetId: string
  ): Promise<boolean> {
    if (!currentUserId) return false;
    const repo = this.dataSource.getRepository(FollowRelation);
    const relation = await repo.findOne({
      where: { userId: currentUserId, targetId },
    });
    return relation !== null;
  }

  /**
   * Check if currentUser has favorited the article.
   */
  private async isFavorited(
    currentUserId: string | undefined,
    articleId: string
  ): Promise<boolean> {
    if (!currentUserId) return false;
    const repo = this.dataSource.getRepository(ArticleFavorite);
    const fav = await repo.findOne({
      where: { userId: currentUserId, articleId },
    });
    return fav !== null;
  }

  /**
   * Get the favorites count for an article.
   */
  private async getFavoritesCount(articleId: string): Promise<number> {
    const repo = this.dataSource.getRepository(ArticleFavorite);
    return repo.count({ where: { articleId } });
  }

  /**
   * Enrich a single article with author, favorited status, and favorites count.
   */
  async enrichArticle(
    article: Article,
    currentUser?: User
  ): Promise<ArticleResponse> {
    const userRepo = this.dataSource.getRepository(User);
    const author = await userRepo.findOneOrFail({ where: { id: article.userId } });
    const favorited = await this.isFavorited(currentUser?.id, article.id);
    const favoritesCount = await this.getFavoritesCount(article.id);
    const following = await this.isFollowing(currentUser?.id, author.id);

    return this.buildArticleResponse(
      article,
      author,
      favorited,
      favoritesCount,
      following
    );
  }

  /**
   * Enrich a list of articles.
   */
  async enrichArticles(
    articles: Article[],
    currentUser?: User
  ): Promise<ArticleResponse[]> {
    return Promise.all(
      articles.map((article) => this.enrichArticle(article, currentUser))
    );
  }

  /**
   * Find article by slug.
   */
  async findBySlug(slug: string): Promise<Article | null> {
    const repo = this.dataSource.getRepository(Article);
    return repo.findOne({ where: { slug }, relations: ["tags"] });
  }

  /**
   * Find article by id.
   */
  async findById(id: string): Promise<Article | null> {
    const repo = this.dataSource.getRepository(Article);
    return repo.findOne({ where: { id }, relations: ["tags"] });
  }

  /**
   * Create a new article.
   */
  async createArticle(
    title: string,
    description: string,
    body: string,
    tagList: string[],
    userId: string
  ): Promise<Article> {
    const repo = this.dataSource.getRepository(Article);
    const tagRepo = this.dataSource.getRepository(Tag);

    // Deduplicate tags
    const uniqueTagNames = [...new Set(tagList)];
    const tags: Tag[] = [];
    for (const name of uniqueTagNames) {
      let tag = await tagRepo.findOne({ where: { name } });
      if (!tag) {
        tag = new Tag(name);
        await tagRepo.save(tag);
      }
      tags.push(tag);
    }

    const article = new Article(title, description, body, [], userId);
    article.tags = tags;
    await repo.save(article);
    return repo.findOneOrFail({ where: { id: article.id }, relations: ["tags"] });
  }

  /**
   * Update an article (only non-empty fields).
   * Mirrors Java Article.update() which uses Util.isEmpty().
   */
  async updateArticle(
    article: Article,
    title?: string,
    description?: string,
    body?: string
  ): Promise<Article> {
    const repo = this.dataSource.getRepository(Article);
    article.update(title, description, body);
    await repo.save(article);
    return repo.findOneOrFail({ where: { id: article.id }, relations: ["tags"] });
  }

  /**
   * Delete an article.
   */
  async deleteArticle(article: Article): Promise<void> {
    const repo = this.dataSource.getRepository(Article);
    await repo.remove(article);
  }

  /**
   * List articles with optional filters: tag, author, favorited.
   * Mirrors Java ArticleQueryService.findRecentArticles().
   */
  async listArticles(
    offset: number,
    limit: number,
    tag?: string,
    author?: string,
    favoritedBy?: string
  ): Promise<{ articles: Article[]; count: number }> {
    const repo = this.dataSource.getRepository(Article);
    const userRepo = this.dataSource.getRepository(User);

    let qb = repo
      .createQueryBuilder("article")
      .leftJoinAndSelect("article.tags", "tag");

    if (tag) {
      qb = qb
        .innerJoin("article.tags", "filterTag", "filterTag.name = :tagName", {
          tagName: tag,
        });
    }

    if (author) {
      const authorUser = await userRepo.findOne({
        where: { username: author },
      });
      if (authorUser) {
        qb = qb.andWhere("article.userId = :authorId", {
          authorId: authorUser.id,
        });
      } else {
        return { articles: [], count: 0 };
      }
    }

    if (favoritedBy) {
      const favUser = await userRepo.findOne({
        where: { username: favoritedBy },
      });
      if (favUser) {
        const favRepo = this.dataSource.getRepository(ArticleFavorite);
        const favs = await favRepo.find({ where: { userId: favUser.id } });
        const favArticleIds = favs.map((f) => f.articleId);
        if (favArticleIds.length === 0) {
          return { articles: [], count: 0 };
        }
        qb = qb.andWhere("article.id IN (:...favArticleIds)", {
          favArticleIds,
        });
      } else {
        return { articles: [], count: 0 };
      }
    }

    const count = await qb.getCount();

    const articles = await qb
      .orderBy("article.createdAt", "DESC")
      .skip(offset)
      .take(limit)
      .getMany();

    return { articles, count };
  }

  /**
   * Get feed articles — articles from users that currentUser follows.
   * Mirrors Java ArticleQueryService.findUserFeed().
   */
  async getFeed(
    userId: string,
    offset: number,
    limit: number
  ): Promise<{ articles: Article[]; count: number }> {
    const followRepo = this.dataSource.getRepository(FollowRelation);
    const articleRepo = this.dataSource.getRepository(Article);

    const followedRelations = await followRepo.find({
      where: { userId },
    });
    const followedUserIds = followedRelations.map((r) => r.targetId);

    if (followedUserIds.length === 0) {
      return { articles: [], count: 0 };
    }

    const count = await articleRepo.count({
      where: { userId: In(followedUserIds) },
    });

    const articles = await articleRepo.find({
      where: { userId: In(followedUserIds) },
      relations: ["tags"],
      order: { createdAt: "DESC" },
      skip: offset,
      take: limit,
    });

    return { articles, count };
  }
}

File: python-fastapi/app/services/article_service.py
import re
import uuid
from datetime import datetime, timezone
from typing import Optional

from sqlalchemy import func
from sqlalchemy.orm import Session

from app.models.models import Article, Tag, User, article_favorites, article_tags, follows
from app.schemas.article import ArticleResponse, ProfileResponse


def to_slug(title: str) -> str:
    return re.sub(r"[&\u{fe30}-\u{ffa0}'\"\s?,\.]+", "-", title.lower())


def _slugify(title: str) -> str:
    slug = re.sub(r"[&'\"\s?,\.]+", "-", title.lower())
    slug = re.sub(r"-+", "-", slug).strip("-")
    return slug


def _build_profile(author: User, current_user: Optional[User], db: Session) -> ProfileResponse:
    is_following = False
    if current_user is not None:
        row = (
            db.query(follows)
            .filter(follows.c.user_id == current_user.id, follows.c.follow_id == author.id)
            .first()
        )
        is_following = row is not None
    return ProfileResponse(
        username=author.username,
        bio=author.bio,
        image=author.image,
        following=is_following,
    )


def _build_article_response(
    article: Article, current_user: Optional[User], db: Session
) -> ArticleResponse:
    tag_names = sorted([t.name for t in article.tags])
    fav_count = (
        db.query(func.count())
        .select_from(article_favorites)
        .filter(article_favorites.c.article_id == article.id)
        .scalar()
    )
    is_favorited = False
    if current_user is not None:
        fav_row = (
            db.query(article_favorites)
            .filter(
                article_favorites.c.article_id == article.id,
                article_favorites.c.user_id == current_user.id,
            )
            .first()
        )
        is_favorited = fav_row is not None

    profile = _build_profile(article.author, current_user, db)

    return ArticleResponse(
        id=article.id,
        slug=article.slug,
        title=article.title,
        description=article.description,
        body=article.body,
        favorited=is_favorited,
        favoritesCount=fav_count or 0,
        createdAt=article.created_at,
        updatedAt=article.updated_at,
        tagList=tag_names,
        author=profile,
    )


def get_article_by_slug(
    slug: str, current_user: Optional[User], db: Session
) -> Optional[ArticleResponse]:
    article = db.query(Article).filter(Article.slug == slug).first()
    if article is None:
        return None
    return _build_article_response(article, current_user, db)


def list_articles(
    db: Session,
    current_user: Optional[User],
    tag: Optional[str] = None,
    author: Optional[str] = None,
    favorited: Optional[str] = None,
    offset: int = 0,
    limit: int = 20,
) -> tuple[list[ArticleResponse], int]:
    limit = min(limit, 100)
    query = db.query(Article)

    if tag:
        query = query.filter(
            Article.id.in_(
                db.query(article_tags.c.article_id)
                .join(Tag, Tag.id == article_tags.c.tag_id)
                .filter(Tag.name == tag)
            )
        )
    if author:
        query = query.filter(Article.user_id.in_(db.query(User.id).filter(User.username == author)))
    if favorited:
        query = query.filter(
            Article.id.in_(
                db.query(article_favorites.c.article_id).filter(
                    article_favorites.c.user_id.in_(
                        db.query(User.id).filter(User.username == favorited)
                    )
                )
            )
        )

    total = query.count()
    articles = query.order_by(Article.created_at.desc()).offset(offset).limit(limit).all()
    return [_build_article_response(a, current_user, db) for a in articles], total


def get_feed(
    db: Session, current_user: User, offset: int = 0, limit: int = 20
) -> tuple[list[ArticleResponse], int]:
    limit = min(limit, 100)
    followed_ids = (
        db.query(follows.c.follow_id).filter(follows.c.user_id == current_user.id).subquery()
    )
    query = db.query(Article).filter(Article.user_id.in_(db.query(followed_ids)))
    total = query.count()
    articles = query.order_by(Article.created_at.desc()).offset(offset).limit(limit).all()
    return [_build_article_response(a, current_user, db) for a in articles], total


def create_article(
    db: Session,
    current_user: User,
    title: str,
    description: str,
    body: str,
    tag_list: Optional[list[str]] = None,
) -> ArticleResponse:
    slug = _slugify(title)
    existing = db.query(Article).filter(Article.slug == slug).first()
    if existing:
        raise ValueError(f"Article with slug '{slug}' already exists")

    now = datetime.now(timezone.utc)
    article = Article(
        id=str(uuid.uuid4()),
        user_id=current_user.id,
        slug=slug,
        title=title,
        description=description,
        body=body,
        created_at=now,
        updated_at=now,
    )

    if tag_list:
        for tag_name in set(tag_list):
            existing_tag = db.query(Tag).filter(Tag.name == tag_name).first()
            if existing_tag:
                article.tags.append(existing_tag)
            else:
                new_tag = Tag(id=str(uuid.uuid4()), name=tag_name)
                db.add(new_tag)
                article.tags.append(new_tag)

    db.add(article)
    db.commit()
    db.refresh(article)
    return _build_article_response(article, current_user, db)


def update_article(
    db: Session,
    article: Article,
    current_user: User,
    title: Optional[str] = "",
    description: Optional[str] = "",
    body: Optional[str] = "",
) -> ArticleResponse:
    if title:
        article.title = title
        article.slug = _slugify(title)
        article.updated_at = datetime.now(timezone.utc)
    if description:
        article.description = description
        article.updated_at = datetime.now(timezone.utc)
    if body:
        article.body = body
        article.updated_at = datetime.now(timezone.utc)

    db.commit()
    db.refresh(article)
    return _build_article_response(article, current_user, db)


def delete_article(db: Session, article: Article) -> None:
    db.delete(article)
    db.commit()
